### ISSUE TITLE
Fullscreen applications doesn't work

### ISSUE THREAD

**Issue by musjj** [awesomeWM/awesome#3844]
<details>
<summary>Output of `awesome --version`:</summary>

```
awesome 4.3 (Too long)
 • Compiled against Lua 5.1.4 (running with LuaJIT 2.1.0-beta3)
 • API level: 4
 • D-Bus support: yes
 • xcb-errors support: no
 • execinfo support: yes
 • xcb-randr version: 1.6
 • LGI version: module 'lgi.version' not found:
	no field package.preload['lgi.version']
	no file './share/lua/5.1/lgi/version.lua'
	no file './lgi/version.lua'
	no file './lgi/version/init.lua'
	no file './lib/lua/5.1/lgi/version.so'
	no file './lgi/version.so'
	no file './lib/lua/5.1/loadall.so'
	no file './lib/lua/5.1/lgi.so'
	no file './lgi.so'
	no file './lib/lua/5.1/loadall.so'
 • Transparency enabled: yes
 • Custom search paths: yes
```
</details>

**How to reproduce the issue:**

1. Set your layout mode to `tile`
2. Open a new terminal window, make sure that it's tiled
3. Run `awesome-client <<<"client.focus.fullscreen = true"`

**Actual result:**

Fullscreen doesn't work. If you set the window to floating mode before fullscreening, it will work just fine:

```bash
awesome-client <<<"client.focus.floating = true; client.focus.fullscreen = true"
```

**Expected result:**

Fullscreen should just work.

**Comment by maverick1872** [comment]
I'm also experiencing the same issue with the latest update.

**Comment by Aire-One** [comment]
Hello,

Can you confirm this issue also happens with the default `awesomerc.lua` and the `Super+f` keybinding.

Also, can you try

```
awesome-client 'client.focus and client.focus.fullscreen = true or require("naughty"). notification { text = "no focused client" }'
```

**Comment by musjj** [comment]
Interesting, when the user config (`~/.config/awesome`) doesn't exist, fullscreen works just fine. I'm not sure what could possible be the cause, since my config is still relatively small. Gonna dig through it a bit.

**Comment by musjj** [comment]
Okay, I think my mistake was that I used the wrong default config (copy-pasted from the stable documentation) as my base config. I rebuilt my config based off the `awesomerc.lua` from the latest commit and I no longer have this problem. Not sure what was the exact cause though.

**Comment by actionless** [comment]
@musjj i having a similar problem on (only) one of my machines, the fastest one. i suspect there is some race condition in handling of client signals or states

**Comment by actionless** [comment]
btw does it happen for you only with luajit?

**Comment by actionless** [comment]
i think it's not possible

i usually just renaming rc.lua in such case

you also could use `awesome --replace --config` flags, but that wouldn't work in my case because i'm using custom startup wrapper for WMs: https://github.com/actionless/awesome_config/blob/master/packaging/awesome_no_argb



btw i checked regarding luajit - it seems to be not the case

so, let me know regarding your cpu model, to get potentially more info

**Comment by Pikabyte** [comment]
Hey, I just started having this problem as well. I think I realized the cause as well. It seems you can no longer fullscreen applications in tiling mode. In floating it works just fine. The issue seems to be caused by this commit:

https://github.com/awesomeWM/awesome/commit/28381d2792856419d4860dd6b9ed9b94b457c7f2

**Comment by actionless** [comment]
oh, that's quite interesting - i double-checked the patch - and the fix is indeed correct, just the initial code was broken initially - it was trying to check if client is floating - and if it's not floating - block geometry request

however that check was always broken (and always passing because check `(not Object1) == Object2` was never happening) - so geometry change was working

so i think the right thing would be to remove that check for floating at all, or modify maximize and fullscreen handlers to not use geometry permission

**Comment by actionless** [comment]
thanks for helping to debug @Pikabyte i found the correct fix

**Comment by Pikabyte** [comment]
> oh, that's quite interesting - i double-checked the patch - and the fix is indeed correct, just the initial code was broken initially - it was trying to check if client is floating - and if it's not floating - block geometry request
> 
> however that check was always broken (and always passing because check `(not Object1) == Object2` was never happening) - so geometry change was working
> 
> so i think the right thing would be to remove that check for floating at all, or modify maximize and fullscreen handlers to not use geometry permission

Yep, just tested the commit before that patch. It works fine there.

**Comment by actionless** [comment]
@Pikabyte you can also test the final code in the PR above ⬆️

**Comment by Pikabyte** [comment]
> @Pikabyte you can also test the final code in the PR above ⬆️
It does seem to work!

**Comment by musjj** [comment]
I tried the patch, unfortunately it didn't work. I used the default `/etc/xdg/awesome/rc.lua` with this snippet embedded to the end:

```lua
client.connect_signal("property::floating", function(c)
  if c.floating then
    c.ontop = true
  else
    c.ontop = false
  end
end)
```

But it still didn't work. This is my hardware, if it helps:

```
OS: NixOS 23.11.20230812.31a86dc (Tapir) x86_64 
Host: LENOVO SHARKBAY 
Kernel: 6.1.45 
Shell: zsh 5.9 
WM: awesome 
CPU: Intel i5-4430 (4) @ 3.200GHz 
GPU: NVIDIA GeForce GT 635 
Memory: 3917MiB / 15957MiB 
```

**Comment by actionless** [comment]
@musjj  try adding print()-s to that function to debug which code branches are getting executed and which aren't to understand why it's not happening - at this point i can't reproduce it anymore, so it's all on you

**Comment by musjj** [comment]
It looks like that both branches gets executed, when you activate fullscreen. I'm not really sure why it happens though. I tried adding a guard like this:

```lua
if c.maximized then
  return
end
```

But both branches still gets executed. 

> at this point i can't reproduce it anymore

Did you try my snippet?

**Comment by musjj** [comment]
Solved it, the correct guard is:

```lua
if c.maximized or c.fullscreen then
  return
end
```

**Comment by actionless** [comment]
oh, apparently that's this - https://github.com/awesomeWM/awesome/issues/2698#issuecomment-466797348

you could also check c.floating itself to be either `false` or `nil`

**Comment by actionless** [comment]
> Did you try my snippet?

i thought i had the same in my config, but it was slightly different when i looked for it:

https://github.com/actionless/awesome_config/blob/5c2acc231681caac4d9508fc242019683045cbc2/config/rules.lua#L92-L97

**Comment by musjj** [comment]
Right, I tried to check the values for `floating` more explicitly, but it didn't work out, one of the branches still gets triggered. 

It looks like that messing with the client options during `fullscreen` can *really* mess things up, including freezing the WM. I had another `arrange` signal that changes the window border width, and I also had to add a `c.fullscreen == true` guard there too to avoid a freeze.

> it was slightly different 

Yeah, a rule won't work for my case, because I want it to apply dynamically. I think rules only apply when the window first spawned.

**Comment by actionless** [comment]
what about checking `_implicitly_floating` property then?

**Comment by Aire-One** [comment]
> @actionless will test for myself later today/tomorrow. Out of curiosity I started looking at the possibility for adding a keybind that would restart awesome with the fallback config explicitly.
> 
> Are you aware if this is achievable via a keybind or not?
> 
> _Edit:_ Based on the [current fn definition](https://github.com/awesomeWM/awesome/blob/28381d2792856419d4860dd6b9ed9b94b457c7f2/awesome.c#L540C1-L545) for restart I presume there is no currently exposed ability for this. Would there be interest in this?

I use LightDM, and have multiple session to start Awesome with different parameters/config/versions.

For example, I have this config to start Awesome from my local git clone 

```
cat /usr/share/xsessions/awesome-git.desktop
[Desktop Entry]
Name=awesome-git
Comment=Highly configurable framework window manager
TryExec=awesome
Exec=/home/aireone/documents/prog/awesome/build/awesome --search /home/aireone/documents/prog/awesome/build/lib --config /home/aireone/documents/prog/awesome/awesomerc.lua
Type=Application
```

**Comment by maverick1872** [comment]
> > @actionless will test for myself later today/tomorrow. Out of curiosity I started looking at the possibility for adding a keybind that would restart awesome with the fallback config explicitly.
> > Are you aware if this is achievable via a keybind or not?
> > _Edit:_ Based on the [current fn definition](https://github.com/awesomeWM/awesome/blob/28381d2792856419d4860dd6b9ed9b94b457c7f2/awesome.c#L540C1-L545) for restart I presume there is no currently exposed ability for this. Would there be interest in this?
> 
> I use LightDM, and have multiple session to start Awesome with different parameters/config/versions.
> 
> For example, I have this config to start Awesome from my local git clone
> 
> ```
> cat /usr/share/xsessions/awesome-git.desktop
> [Desktop Entry]
> Name=awesome-git
> Comment=Highly configurable framework window manager
> TryExec=awesome
> Exec=/home/aireone/documents/prog/awesome/build/awesome --search /home/aireone/documents/prog/awesome/build/lib --config /home/aireone/documents/prog/awesome/awesomerc.lua
> Type=Application
> ```

Yea that's one manner of doing it but requires a logout/login. Which is what I was trying to avoid. Ended up just defining a pretty simple shell alias to ease the process without relogging myself. 

```zsh
swap-awesome-config() {
  swap ~/.config/awesome/rc.lua ~/.config/awesome/rc.lua.inactive
  echo 'awesome.restart()' | awesome-client
}
```

**Comment by actionless** [comment]
damn it, stop mentioning me when quoting
